A function to find blocks suitable to execute on GPU

daviditen opened the pull request:
Add a function that returns `true` if the only calls in a block are
`FAST_AND_LOCAL` primitives according to the analysis implemented in
the `optimizeOnClauses` pass. It can also return `true` for blocks
containing calls to functions that only use `FAST_AND_LOCAL` primitives
recursively, but this is currently disabled with the `allowFnCalls=false`
argument.

Call this function in the GPU outliner code in addition to the check it was
already doing. We can remove the other check in another step.

e-kayrakli commented on compiler/optimizations/deadCodeElimination.cpp:
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
+  return true;
+}
+
+static void markGPUSuitableLoops() {
```
This is currently only for debugging, right? I think we should either:

1. keep this, but put an if around its call site that's `false` by default
2. remove it altogether

We should probably do 1 for the time being as we need as much debugging support as we can until we stabilize the implementation.

e-kayrakli replied:
(We could probably use `debugPrint` at the callsite)

e-kayrakli commented on compiler/optimizations/deadCodeElimination.cpp:
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
+      // only primitives that are fast and local are allowed for now
+      bool inLocal = inLocalBlock(call);
+      int is = classifyPrimitive(call, inLocal);
+      if ((is != FAST_AND_LOCAL)) {
```
I think relying on `FAST_AND_LOCAL` is good enough for now. But I think we'll have to do our own classification for GPU. I am not sure if there's a strict relation between the ability of running a primitive on an AM handler (`FAST`) vs running it on GPU.

e-kayrakli commented on compiler/optimizations/deadCodeElimination.cpp:
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
+
+bool debugPrint = false; 
+bool allowFnCallsFromGPU = false;
+int indent = 0;
```
Can you rename this to have GPU in the name?

e-kayrakli commented on compiler/optimizations/deadCodeElimination.cpp:
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
+extern int classifyPrimitive(CallExpr *call, bool inLocal);
+extern bool inLocalBlock(CallExpr *call);
+
+bool debugPrint = false; 
```
Can you rename this to have GPU in the name?

e-kayrakli commented on compiler/optimizations/deadCodeElimination.cpp:
```diff
@@ -26,6 +26,7 @@
 #include "CForLoop.h"
 #include "driver.h"
 #include "expr.h"
+#include "ForallStmt.h"
```
We *probably* don't need this. But I'm not sure.

e-kayrakli commented on compiler/include/ForallStmt.h:
```diff
@@ -83,6 +83,9 @@ class ForallStmt final : public Stmt
   CallExpr* zipCall() const;
 
+  bool      isGPUSuitable() const; // is suitable for running on GPU
+  void      setIsGPUSuitable(bool isSuitable); // if passed suitability checks
```
Do we use these two?

e-kayrakli reviewed (commented):
Thanks! I think my main worry is that we are relying on the fast/local primitive classification for the `on` statements here. But in the short term, I'm OK with it as long as this PR passes a non-gpu gasnet paratest. Besides that, I have some more nitpicky inline comments.

Could you also reflow the code to fit into 80 columns?

daviditen commented:
I think I've addressed your feedback now.

I was able to remove the primitives that were previously added to `classifyPrimitive` and passed a full paratest without them.

e-kayrakli commented on compiler/include/ForallStmt.h:
```diff
@@ -83,6 +83,7 @@ class ForallStmt final : public Stmt
   LabelSymbol* continueLabel();      // create it if not already
   CallExpr* zipCall() const;
 
+
```
extra whitespace changes in this file

e-kayrakli commented on compiler/passes/checkResolved.cpp:
```diff
@@ -123,6 +122,7 @@ checkResolved() {
       }
     }
   }
+
```
extra whitespace changes in this file

e-kayrakli reviewed (approved):
I noted two files that have some leftover whitespace changes. Besides that all's good by me!
